cloud/misc/ovirt.py

```diff
@@ -199,231 +285,944 @@
-    vmcpus        = module.params['instance_cpus']      # number of cpu
-    vmnic         = module.params['instance_nic']       # network interface
-    vmnetwork     = module.params['instance_network']   # logical network
-    vmmem         = module.params['instance_mem']       # mem size 
-    vmdisk_alloc  = module.params['disk_alloc']         # thin, preallocated
-    vmdisk_int    = module.params['disk_int']           # disk interface virtio or ide
-    vmos          = module.params['instance_os']        # Operating System
-    vmtype        = module.params['instance_type']      # server or desktop
-    vmcores       = module.params['instance_cores']     # number of cores
-    sdomain       = module.params['sdomain']            # storage domain to store disk on
-    region        = module.params['region']             # oVirt Datacenter
-    #initialize connection
-    c = conn(url+"/api", user, password)
-
-    if state == 'present':
-        if get_vm(c, vmname) == "empty":
-            if resource_type == 'template':
-                create_vm_template(c, vmname, image, zone)
-                module.exit_json(changed=True, msg="deployed VM %s from template %s"  % (vmname,image))
-            elif resource_type == 'new':
-                # FIXME: refactor, use keyword args.
-                create_vm(c, vmtype, vmname, zone, vmdisk_size, vmcpus, vmnic, vmnetwork, vmmem, vmdisk_alloc, sdomain, vmcores, vmos, vmdisk_int)
-                module.exit_json(changed=True, msg="deployed VM %s from scratch"  % vmname)
+    # noinspection PyBroadException
+    def add_tags(self):
+        """
+        :rtype: list[ovirtsdk.infrastructure.brokers.Tag]
+        """
+        tags = self.module.params['tags']
+
+        if tags:
+            new_tags = set([tag.strip().lower() for tag in tags.strip().replace(" ", "_").split(',')])
+
+            try:
+                existing_tags = set([tag.get_name() for tag in self.conn.tags.list()])
+            except Exception:
+                self.module.fail_json(msg=u"Could not get existing tags from server")
             else:
-                module.exit_json(changed=False, msg="You did not specify a resource type")
-        else:
-            module.exit_json(changed=False, msg="VM %s already exists" % vmname)
+                tags_to_add = []
+                for tag in new_tags:
+                    if tag not in existing_tags:
+                        try:
+                            self.conn.tags.add(params.Tag(name=tag))
+                        except:
+                            self.module.fail_json(msg=u"Failed to add tag '{}' to ovirt".format(tag))
+                    try:
+                        infra_tag = self.conn.tags.get(name=tag)
+                        tags_to_add.append(infra_tag)
+                    except:
+                        self.module.fail_json(msg=u"Failed to get tag '{}' from ovirt".format(tag))
+                return tags_to_add
```

Comment:
You should "return []" outside the if, to cover the case where tags = '' (the default, that evaluates to False in python)